safe_transaction_service/history/services/collectibles_service.py

```diff
@@ -424,78 +444,82 @@ def get_token_uris(
+                "Getting token uris from blockchain for %d addresses with tokenIds",
+                len(not_found_uris),
             )
             blockchain_token_uris = {
                 address_with_token_id: token_uri if token_uri else None
                 for address_with_token_id, token_uri in zip(
-                    not_found_cache,
-                    self.ethereum_client.erc721.get_token_uris(not_found_cache),
+                    not_found_uris,
+                    self.ethereum_client.erc721.get_token_uris(not_found_uris),
                 )
             }
-            logger.debug("Got token uris")
-            if blockchain_token_uris:
-                self.cache_token_uri.update(blockchain_token_uris)
-                pipe = self.redis.pipeline()
-                redis_map_to_store = {
-                    get_redis_key(address_with_token_id): token_uri
-                    if token_uri is not None
-                    else ""
-                    for address_with_token_id, token_uri in blockchain_token_uris.items()
-                }
-                pipe.mset(redis_map_to_store)
-                for key in redis_map_to_store.keys():
-                    pipe.expire(key, 60 * 60 * 24)  # 1 day of caching
-                pipe.execute()
-        except (IOError, ValueError) as exc:
-            raise NodeConnectionException from exc
+            logger.debug("Got token uris from blockchain")
+        except (IOError, ValueError):
+            logger.warning(
+                "Problem when getting token uris from blockchain, trying individually",
+                exc_info=True,
+            )
+            blockchain_token_uris = {}
+            for not_found_uri in not_found_uris:
+                try:
+                    token_uri = self.ethereum_client.erc721.get_token_uris(
+                        [not_found_uri]
+                    )[0]
+                    blockchain_token_uris[not_found_uri] = (
+                        token_uri if token_uri else None
+                    )
+                except ValueError:
+                    blockchain_token_uris[not_found_uri] = None
+                    logger.warning(
+                        "ValueError when getting token uri from blockchain for token and tokenId %s",
+                        not_found_uri,
+                        exc_info=True,
+                    )
+                except IOError as exc:
+                    raise NodeConnectionException from exc
+
+        if blockchain_token_uris:
+            pipe = self.redis.pipeline()
+            redis_map_to_store = {
+                get_redis_key(address_with_token_id): token_uri
+                if token_uri is not None
+                else ""
+                for address_with_token_id, token_uri in blockchain_token_uris.items()
+            }
+            pipe.mset(redis_map_to_store)
+            for key in redis_map_to_store.keys():
+                pipe.expire(key, 60 * 60 * 24)  # 1 day of caching
```

Comment:
Should this be an env var?

Reply:
I will change this in another PR, as there are some caching times that can be also configured